sdk/client-core/core/src/main/java/io/clientcore/core/http/pipeline/HttpPipelinePolicy.java

```diff
@@ -5,21 +5,60 @@
 
 import io.clientcore.core.http.models.HttpRequest;
 import io.clientcore.core.http.models.Response;
+import io.clientcore.core.util.ClientLogger;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
 
 /**
  * A policy within the {@link HttpPipeline}.
  *
  * @see HttpPipeline
  */
-@FunctionalInterface
-public interface HttpPipelinePolicy {
+public abstract class HttpPipelinePolicy implements Cloneable {
+    private static final ClientLogger LOGGER = new ClientLogger(HttpPipelinePolicy.class);
+    private HttpPipelinePolicy nextPolicy;
+
     /**
-     * Processes the provided HTTP request and invokes the next policy synchronously.
+     * Processes an HTTP request and invokes the next policy in the chain synchronously.
      *
      * @param httpRequest The HTTP request.
-     * @param next The next policy to invoke.
+     * @param httpPipeline The HTTP pipeline the request will be sent through.
+     *
+     * @return A response produced from sending the HTTP request.
+     */
+    public Response<?> process(HttpRequest httpRequest, HttpPipeline httpPipeline) {
```

Comment:
If we're going to be changing the API to remove `HttpPipelineNextPolicy` I think we should do a major break here instead of trying to retain a similar flow as `azure-core`. Right now, each pipeline has a mutable state with `nextPolicy` and `HttpPipeline` is supposed to be thread-safe which this will break. If two requests are being made where a common `RetryPolicy` is shared between two pipelines the `nextPolicy` may not point to the right following policy.

What I'll propose, split `process` into two methods, `onRequest(HttpRequest) : HttpRequest` and `onResponse(HttpResponse) : HttpResponse`. This makes `HttpPipelinePolicy` a stateless call and removes, what I believe wasn't a great pattern before, where the pipeline policy was both a mutator of requests and responses and managed a small bit of state management as it was forced to call `HttpPipelineNextPolicy.process`, and if it didn't now it broke the `HttpPipeline` call.

Simplifying `HttpPipelinePolicy` to `onRequest` and `onResponse` removes state management. It no longer needs to make a call to process the next policy, allowing the `HttpPipelinePolicy` to be a request and response mutator and `HttpPipeline` completely owns the state management. This would also allow for policies to be extended into more responsibilities in the future such as `onException`, if we find additional requirements (though I believe `onRequest` and `onResponse` would be sufficient for a long time).